### PR TITLE
Add whisper-stream-pcm: stdin/pipe PCM streaming with VAD support

### PR DESCRIPTION
**Pull request by rmorse**
## Summary

- Adds `whisper-stream-pcm`, a new example binary for streaming raw PCM audio via stdin/pipe for real-time transcription — no WAV headers or SDL2 required
- Supports both simple (energy-based) and Silero VAD for speech segmentation
- Accepts 32-bit float or 16-bit signed integer PCM (16kHz, mono)

## Usage

```bash
whisper-stream-pcm -m models/ggml-base.en.bin
```

With Silero VAD:
```bash
whisper-stream-pcm -m models/ggml-base.en.bin --vad --vad-model models/silero-vad.bin
```

Using FFMPEG to test/pipe in the data
```bash
ffmpeg -i audio.wav -f f32le -ar 16000 -ac 1 pipe:1 | whisper-stream-pcm -m models/ggml-base.en.bin --vad --vad-model models/silero-vad.bin
```

## Changes

- `examples/stream-pcm/` — new example binary (stream-pcm.cpp, CMakeLists.txt, README.md)
- `examples/CMakeLists.txt` — register stream-pcm example
- `tests/test-stream-pcm.cpp` — test for stdin/pipe PCM streaming
- `tests/CMakeLists.txt` — register test
- `README.md` — mention stream-pcm in examples list

## Test plan

- [ ] `whisper-stream-pcm` builds on Linux/macOS/Windows
- [x] Piping PCM audio via ffmpeg produces transcription output
- [x] `--vad` (simple) and `--vad --vad-model` (Silero) both work correctly
- [x] `ctest` passes test-stream-pcm

### CODE REVIEW

**Comment by rmorse** [comment]
@ggerganov I just merged with 1.8.6 and this is still looking good:
https://github.com/rmorse/whisper.cpp/releases/tag/v1.8.6-stream-pcm

**Comment by ggerganov** [comment]
Alright thanks. We'll take a look. cc @danbev

File: examples/stream-pcm/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+```bash
+mkfifo /tmp/whisper.pcm
+./build/bin/whisper-stream-pcm -m ./models/ggml-base.en.bin --input /tmp/whisper.pcm --format s16 --sample-rate 16000 --step 1000 --length 10000 --keep 500
+```
+
+Example of piping a WAV file using ffmpeg (optional, `-re` for realtime pacing):
+
+```bash
+ffmpeg -re -i samples/jfk.wav -f s16le -ac 1 -ar 16000 - | \
```
Comment (danbev) [inline examples/stream-pcm/README.md:30]: Perhaps adding `-v quiet` would be nice avoid the ffmpeg status log messages. For example,  this is what the output look at the moment:
```console
[Start streaming]
 And so my fellow America!size=      78KiB time=00:00:02.56 bitrate= 249.6kbits/s speed=1.27x elapsed=0:00
 And so, my fellow Americans!size=     110KiB time=00:00:03.58 bitrate= 251.4kbits/s speed=1.18x elapsed=0
 And so my fellow Americans askedsize=     140KiB time=00:00:04.54 bitrate= 252.4kbits/s speed=1.13x elaps
 And so my fellow Americans, ask not!size=     172KiB time=00:00:05.56 bitrate= 253.1kbits/s speed= 1.1x e
 And so my fellow Americans ask not what you are coming to.size=     204KiB time=00:00:06.59 bitrate= 253.
 And so my fellow Americans ask not what your country can do for you.size=     236KiB time=00:00:07.61 bit
 And so my fellow Americans, ask not what your country can do for you.size=     266KiB time=00:00:08.57 bi
 And so my fellow Americans, ask not what your country can do for you, ask what you...
 what you can do for yoursize=     330KiB time=00:00:10.62 bitrate= 254.5kbits/s speed=1.05x elapsed=0:00:[out#0/s16le @ 0x6000012d0180] video:0KiB audio:344KiB subtitle:0KiB other streams:0KiB global headers:0KiB muxing overhead: 0.000000%
size=     344KiB time=00:00:11.00 bitrate= 256.0kbits/s speed=1.05x elapsed=0:00:10.44
 what you can do for your country.
```
With the `-v quiet` flag:
```console
ffmpeg -v quiet -re -i samples/jfk.wav -f s16le -ac 1 -ar 16000 - | ./build-stream-pcm/bin/whisper-stream-pcm -m ./models/ggml-base.en.bin --format s16 --sample-rate 16000 --step 1000 --length 10000 --keep 500
whisper_init_from_file_with_params_no_state: loading model from './models/ggml-base.en.bin'
...
[Start streaming]
 And so my fellow Americans, ask not what your country can do for you, ask what you...
 what you can do for your country.
```

File: examples/stream-pcm/stream-pcm.cpp [hunk 2]
```diff
@@ -0,0 +1,901 @@
...
+                m_thread.detach();
+            }
+        }
+
+        m_running = false;
+
+        return true;
+    }
+
+    bool clear() {
```
Comment (danbev) [inline examples/stream-pcm/stream-pcm.cpp:245]: This function does not seem to be called and it does not seem to have any failure conditions (it always returns true). Perhaps removing or changing the return type to void should be considered.

File: examples/stream-pcm/stream-pcm.cpp [hunk 3]
```diff
@@ -0,0 +1,901 @@
...
+            m_in = fopen(input_path.c_str(), "rb");
+            m_owns_input = true;
+        }
+
+        if (!m_in) {
+            fprintf(stderr, "%s: failed to open input '%s'\n", __func__, input_path.c_str());
+            return false;
+        }
+
+        m_audio.resize((m_sample_rate*m_len_ms)/1000);
```
Comment (danbev) [inline examples/stream-pcm/stream-pcm.cpp:195]: Nit: For consistency: 
```suggestion
m_audio.resize((m_sample_rate * m_len_ms) / 1000);

File: examples/stream-pcm/stream-pcm.cpp [hunk 4]
```diff
@@ -0,0 +1,901 @@
...
+                const size_t drop = n_samples - (m_audio.size() - m_audio_len);
+                m_audio_read = (m_audio_read + drop) % m_audio.size();
+                m_audio_len -= drop;
+            }
+
+            if (m_audio_pos + n_samples > m_audio.size()) {
+                const size_t n0 = m_audio.size() - m_audio_pos;
+                memcpy(&m_audio[m_audio_pos], data, n0 * sizeof(float));
+                memcpy(&m_audio[0], data + n0, (n_samples - n0) * sizeof(float));
+        } else {
```
Comment (danbev) [inline examples/stream-pcm/stream-pcm.cpp:403]: Nit: Indentation is slightly off here.

File: examples/stream-pcm/stream-pcm.cpp [hunk 5]
```diff
@@ -0,0 +1,901 @@
...
+        }
+
+        result.resize(n_samples);
+
+        int s0 = (int) m_audio_pos - (int) n_samples;
+        if (s0 < 0) {
+            s0 += (int) m_audio.size();
+        }
+
+        if (s0 + (int) n_samples > (int) m_audio.size()) {
```
Comment (danbev) [inline examples/stream-pcm/stream-pcm.cpp:274]: Perhaps this could be extracted into a `ring_read` function and then reused in [pop_ms](https://github.com/ggml-org/whisper.cpp/pull/3653/changes#diff-bd879c9752d9925d605d4ceabecbb5c6c0f698a96ca14b625133ca8f24dfca55R303-R310).

File: examples/stream-pcm/stream-pcm.cpp [hunk 6]
```diff
@@ -0,0 +1,901 @@
...
+    std::ofstream fout;
+    if (params.fname_out.length() > 0) {
+        fout.open(params.fname_out);
+        if (!fout.is_open()) {
+            fprintf(stderr, "%s: failed to open output file '%s'!\n", __func__, params.fname_out.c_str());
+            return 1;
+        }
+    }
+
+    wav_writer wavWriter;
```
Comment (danbev) [inline examples/stream-pcm/stream-pcm.cpp:593]: Nit: Consider using snake case for here, perhaps `wav_out` or something perhaps.

**Review by danbev — APPROVED** [review]

File: examples/stream-pcm/stream-pcm.cpp (hunk 2 above)
Comment (rmorse) [reply examples/stream-pcm/stream-pcm.cpp]: Yup removed

File: examples/stream-pcm/README.md (hunk 1 above)
Comment (rmorse) [reply examples/stream-pcm/README.md]: So I had another example in the root README using `-loglevel error` rather than `quiet` - updated this one to match - but let me know if you want this switching over to `quiet` in both places.

File: examples/stream-pcm/stream-pcm.cpp (hunk 3 above)
Comment (rmorse) [reply examples/stream-pcm/stream-pcm.cpp]: Done

File: examples/stream-pcm/stream-pcm.cpp (hunk 4 above)
Comment (rmorse) [reply examples/stream-pcm/stream-pcm.cpp]: Done

File: examples/stream-pcm/stream-pcm.cpp (hunk 5 above)
Comment (rmorse) [reply examples/stream-pcm/stream-pcm.cpp]: Extracted ring_read() and used it in pop_ms(). get() turned out to be unused too, also removed.

File: examples/stream-pcm/stream-pcm.cpp (hunk 6 above)
Comment (rmorse) [reply examples/stream-pcm/stream-pcm.cpp]: Renamed to `wav_out`

**Comment by rmorse** [comment]
@danbev & @ggerganov - completed all feedback and only a couple of comments left open.

Also updated and merged with master.

And since the AGENTS protocol came in after I started this PR - I do use AI and on this PR.. but.. on a leash :)
